### PR TITLE
test: honor turn-seam contracts in conversation memory follow-up

### LINKED ISSUE #5198: Tests: honour turn-seam contracts in test_conversation_memory_follow_up.py
## Goal

Make the turn-stage fakes in `tests/core/agent_harness/prompts/test_conversation_memory_follow_up.py` honour the contracts they replace.

Two patterns to remove, both of which keep a test green when the contract
changes underneath it:

- a stage injected as `lambda *_a, **_k` — ignores the callable Protocol
- an answer faked as `type("Run", (), {...})()` — an anonymous object with
  one attribute, standing in for a real result type

## Swallow-all seam stubs in this file (2)

```
  131: answer=lambda *_a, **_k: None,
  132: gather=lambda *_a, **_k: None,
```

Each of these should take the signature its Protocol declares.
`EvidenceGatherer`, for example, is
`__call__(self, text: str, *, turn_plan: Any = None)`
(`core/agent_harness/ports.py`), so its stub is:

```python
def no_evidence(text: str, *, turn_plan: Any = None) -> GatheredEvidence | None:
    """An EvidenceGatherer that finds nothing."""
    return None
```

Add a required keyword to the Protocol and that version fails at once;
the swallow-all lambda keeps passing.

## Order

Depends on #5187 (shared turn-seam helpers). Wait for it to land.

## Acceptance

- [ ] No `lambda *_a, **_k` passed as a stage in `tests/core/agent_harness/prompts/test_conversation_memory_follow_up.py`
- [ ] Stubs are named functions reusing the shared helpers
- [ ] Test intent unchanged: `answer=None` means the real stage runs, a
      no-op stub means no answer — do not swap one for the other
- [ ] `uv run pytest tests/core/agent_harness/prompts/test_conversation_memory_follow_up.py` green, no product change

## Out of scope

Other files, monkeypatch lambdas that are not stage injections, and
product Protocol bodies.

## Guidelines

`AGENTS.md` — Protocols as callable contracts; test fakes are named `def`s.

### PR DESCRIPTION
**Pull request by nightcityblade**
Fixes #5198

#### Describe the changes you have made in this PR -

- Replaced the permissive variadic answer lambda with a named, typed no-answer stub that matches `StreamAnswerFn`.
- Reused the shared `no_evidence` gather helper so future Protocol signature changes are caught by type checking.
- Preserved the test's existing no-op answer and evidence behavior.

### Demo/Screenshot for feature changes and bug fixes -

```text
$ make lint
All checks passed!

$ make format-check
3598 files already formatted

$ make typecheck
Success: no issues found in 1860 source files

$ uv run mypy tests/core/agent_harness/prompts/test_conversation_memory_follow_up.py
Success: no issues found in 1 source file

$ uv run pytest tests/core/agent_harness/prompts/test_conversation_memory_follow_up.py
10 passed in 0.99s
```

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

The test previously injected variadic lambdas that accepted any future call shape, so interface drift would not be caught by static checking. I replaced the answer lambda with a typed function whose parameter names and types match the callable Protocol and reused the shared `no_evidence` helper introduced by #5187 for the gather seam. I considered passing `answer=None`, but that selects the real answer stage and changes the test's intent, so an explicit no-op answer implementation is retained.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests (N/A: this is a test-only contract-hardening change)
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

### CODE REVIEW

File: tests/core/agent_harness/prompts/test_conversation_memory_follow_up.py [hunk 1]
```diff
@@ -2,10 +2,18 @@
...
+from core.agent_harness.spi.accounting import LlmRunInfo
 from core.agent_harness.turns.headless_adapters import InMemorySessionState, NoopTurnAccounting
 from core.agent_harness.turns.orchestrator import run_turn
 from core.agent_harness.turns.turn_results import ToolCallingTurnResult
+from tests.shared.harness_turn_driver import no_evidence
+
+
+def no_answer(text: str, request: AnswerRequest) -> LlmRunInfo | None:
+    """A StreamAnswerFn that produces no answer."""
+    return None
```
[the remark this answers is not in the document]
Comment (nightcityblade) [reply tests/core/agent_harness/prompts/test_conversation_memory_follow_up.py:16]: Thanks for flagging this. I am keeping this one as a named `def` because #5198 explicitly requires these injected stage stubs to be named functions that honor their callable Protocols. Unlike a general trivial callback, the explicit `text: str, request: AnswerRequest` signature is the behavior under test: it makes future `StreamAnswerFn` contract drift fail static checking, which an inline variadic lambda would continue to accept. The issue guidelines specifically call out Protocol test fakes as named defs, so this is an intentional narrow exception to the general trivial-lambda convention.

**Comment by nightcityblade** [comment]
@greptile review

**Comment by Devesh36** [comment]
Good first issues are only for new contributor....closing thiss
